### PR TITLE
Adding KFAC

### PR DESCRIPTION
**Pull request by shaipranesh2**
# Pull Request Template

## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by peastman** [comment]
Since the KFAC code was copied from https://github.com/alecwangcq/KFAC-Pytorch, we should add a license header at the top of the file.  Credit the original author, give a link to the original repository, and state the exact license terms he gave you permission to use it under.

File: deepchem/models/torch_models/torch_model.py [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import numpy as np
-import torch
```
Comment (arunppsg) [inline deepchem/models/torch_models/torch_model.py:2]: We should preserve this `import torch` statement here.

File: deepchem/models/tests/test_optimizers.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
+import deepchem as dc
 import deepchem.models.optimizers as optimizers
 import unittest
+from deepchem.models.torch_models import kfac_optimizer
```
Comment (arunppsg) [inline deepchem/models/tests/test_optimizers.py:4]: This import statement depends on torch module. It is better to keep it's scope local to the test method. In this case, it will `test_KFAC` method

File: deepchem/models/torch_models/torch_model.py [hunk 3]
```diff
@@ -945,10 +953,10 @@ def default_generator(
...
     """
     for epoch in range(epochs):
-      for (X_b, y_b, w_b, ids_b) in dataset.iterbatches(
-          batch_size=self.batch_size,
-          deterministic=deterministic,
-          pad_batches=pad_batches):
+      for (X_b, y_b, w_b,
+           ids_b) in dataset.iterbatches(batch_size=self.batch_size,
+                                         deterministic=deterministic,
+                                         pad_batches=pad_batches):
```
Comment (tonydavis629) [inline deepchem/models/torch_models/torch_model.py:959]: Was this done by yapf? Doesn't seem right. I'm not sure if all these formatting changes are necessary.

File: deepchem/models/tests/test_optimizers.py [hunk 4]
```diff
@@ -1,3 +1,5 @@
+from haiku import Linear
```
Comment (peastman) [inline deepchem/models/tests/test_optimizers.py:1]: Did you intend to add this?  I don't think you use it anywhere.  Since it requires JAX, it should only be imported inside functions that are marked for JAX only.

Comment (shaipranesh2) [reply deepchem/models/tests/test_optimizers.py:1]: yea must have been added by the IDE by default and just slipped in. Will change it

File: deepchem/models/optimizers.py [hunk 5]
```diff
@@ -678,3 +677,44 @@ def _create_jax_schedule(self):
...
+
+class KFAC(Optimizer):
+  """The Second order gradient optimiation algorithm which uses an approximation to calculate the inverse of the Fischer matrrix"""
+
+  def __init__(self, **kwargs):
+    """
+    Parameters:
+    -----------
+    model: torch.nn.Module
+    The model to be optimized.
```
Comment (peastman) [inline deepchem/models/optimizers.py:690]: The description line for each argument needs to be indented.  See https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard.

File: deepchem/models/tests/test_optimizers.py [hunk 6]
```diff
@@ -255,13 +262,70 @@ def test_linearCosine_decay_jax(self):
...
     assert isinstance(jaxopt, optax.GradientTransformation)
+
+  @pytest.mark.torch
+  def test_KFAC(self):
+    """test creating a KFAC optimizer"""
+    import torch
+    import numpy as np
+
+    np.random.seed(123)
+    """linear layers test"""
```
Comment (peastman) [inline deepchem/models/tests/test_optimizers.py:282]: Comments inside functions should start with `#`.  Only docstrings are string literals.

File: deepchem/models/tests/test_optimizers.py [hunk 7]
```diff
@@ -255,13 +262,70 @@ def test_linearCosine_decay_jax(self):
...
+    model = dc.models.TorchModel(pytorch_model,
+                                 dc.models.losses.L2Loss(),
+                                 optimizers=optimizers.KFAC(model=pytorch_model,
+                                                            learning_rate=0.1,
+                                                            Tinv=50))
+    model.fit(dataset, nb_epoch=100)
+
+    scores = model.evaluate(dataset, [metric])
+    assert scores[metric.name] > 0.9
+    """Conv2d layers test"""
```
Comment (peastman) [inline deepchem/models/tests/test_optimizers.py:303]: Same here.

File: deepchem/models/torch_models/kfac_optimizer.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+      self.m_gg[module] *= self.stat_decay + gg * (1 - self.stat_decay)
+
+  def _prepare_model(self):
+    """"
+    Attaches hooks(saving the ouptut and grad according to the update function) to the model for
+    to calculate gradients at every step.
+    """
+    count = 0
+    print(self.model)
+    print("=> We keep following layers in KFAC. ")
```
Comment (peastman) [inline deepchem/models/torch_models/kfac_optimizer.py:257]: Logging should be done with the `logging` package rather than `print()` calls.  But is this really intended to be logged, or is it just debugging code that got left in?

**Comment by peastman** [comment]
Mypy seems to think `Tensor.view` is a Tensor rather than a function that returns a Tensor.  I'm not sure why.

I do understand the errors it's reporting in lines 138-9.  You declare `padding` to have type `Union[str, Tuple[int, ...]]`.  That means it could be either a string or a tuple.  But those lines assume it's a tuple of ints.  Assuming `padding` really can be a `str`, those lines are incorrect.  You need to first check `if isinstance(padding, str)` and handle that case differently.  Or if it really has to be a tuple, then you need to fix the declaration.

**Comment by peastman** [comment]
Possibly the problem is that you're circumventing the type system by passing in a layer name as a string.  Like in these lines:

```python
    if classname == 'Conv2d':
      p_grad_mat = m.weight.grad.data.view(m.weight.grad.data.size(0),
                                           -1)  # n_filters * (in_c * kw * kh)
```

You know that if `classname` is `'Conv2d'`, that means `m` is `torch.nn.Conv2d` object, and therefore it has a field called `weight` which is a Tensor.  But mypy doesn't know that.  And of course, you don't really know it either.  That's how you intend the function to be called, but you can't guarantee it always will be.  What if you eliminated the `classname` argument and changed those lines to

```python
    if isinstance(m, torch.nn.Conv2d):
      p_grad_mat = m.weight.grad.data.view(m.weight.grad.data.size(0),
                                           -1)  # n_filters * (in_c * kw * kh)
```

**Comment by shaipranesh2** [comment]
@peastman @tonydavis629 I have done some changes by adding in isinstance to check the type of the weight.grad.data and bias.grad.data and to proceed further only if they are torch tensor type, otherwise raises an error. 
Also, have done the same for the padding , where it checks for str, int and tuple, and proceeds to add the padding accordingly.

File: deepchem/models/torch_models/kfac_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+    elif isinstance(padding, str):
+      if padding == 'VALID':
+        pad = int((kernel_size[0] - 1) / 2)
+        x = torch.nn.functional.pad(x, (pad, pad, pad, pad)).data
+
+      x = x.unfold(2, kernel_size[0], stride[0])
+      x = x.unfold(3, kernel_size[1], stride[1])
+      x = x.transpose_(1, 2).transpose_(2, 3).contiguous()
+      x = x.view(x.size(0), x.size(1), x.size(2),
+                 x.size(3) * x.size(4) * x.size(5))
```
Comment (peastman) [inline deepchem/models/torch_models/kfac_optimizer.py:164]: Are you sure the indentation here is correct?  These lines are only supposed to be executed when `padding` is a string?

**Review by peastman — APPROVED** [review]
Assuming all the test failures are unrelated (I haven't checked), this looks good.

**Comment by shaipranesh2** [comment]
@rbharath can you merge this

**Review by tonydavis629 — APPROVED** [review]
